pydantic/_internal/_decorators.py

```diff
@@ -208,6 +210,90 @@ def bind_to_cls(self, cls: Any) -> Decorator[DecoratorInfoType]:
         )
 
 
+def get_bases(tp: type[Any]) -> tuple[type[Any], ...]:
+    if is_typeddict(tp):
+        return tp.__orig_bases__  # type: ignore
+    try:
+        return tp.__bases__
+    except AttributeError:
+        return ()
+
+
+def mro(tp: type[Any]) -> tuple[type[Any], ...]:
+    """Calculate the Method Resolution Order of bases using the C3 algorithm.
+
+    See https://www.python.org/download/releases/2.3/mro/
+    """
+
+    # try to use the existing mro, for performance mainly
+    # but also because it helps verify the implementation below
+    if not is_typeddict(tp):
+        try:
+            return tp.__mro__
+        except AttributeError:
+            # GenericAlias and some other cases
+            pass
+
+    def merge_seqs(seqs: list[deque[type[Any]]]) -> Iterable[type[Any]]:
+        while True:
+            non_empty = [seq for seq in seqs if seq]
+            if not non_empty:
+                # Nothing left to process, we're done.
+                return
+            candidate: type[Any] | None = None
+            for seq in non_empty:  # Find merge candidates among seq heads.
+                candidate = seq[0]
+                not_head = [s for s in non_empty if candidate in islice(s, 1, None)]
+                if not_head:
+                    # Reject the candidate.
+                    candidate = None
+                else:
+                    break
+            if not candidate:
+                raise TypeError('Inconsistent hierarchy, no C3 MRO is possible')
+            yield candidate
+            for seq in non_empty:
+                # Remove candidate.
+                if seq[0] == candidate:
+                    seq.popleft()
+
+    bases = get_bases(tp)
+    seqs = [deque(mro(base)) for base in bases] + [deque(bases)]
+    res = tuple(merge_seqs(seqs))
+
+    return (tp,) + res
+
+
+def get_attribute_from_bases(tp: type[Any], name: str) -> Any:
+    """
+    Get the attribute from the next class in the MRO that has it,
+    aiming to simulate calling the method on the actual class.
+
+    The reason for iterating over the mro instead of just getting
+    the attribute (which would do that for us) is to support TypedDict,
+    which lacks a real __mro__, but can have a virtual one constructed
+    from its bases (as done here).
+
+    Args:
+        tp (type[Any]): The type or class to search for the attribute.
+        name (str): The name of the attribute to retrieve.
+
+    Returns:
+        Any: The attribute value, if found.
+
+    Raises:
+        AttributeError: If the attribute is not found in any class in the MRO.
+    """
+    e: Exception | None = None
+    for base in [tp, *reversed(mro(tp))]:
+        try:
+            return getattr(base, name)
```

Comment:
`hasattr` and `if` is faster than catching exception on `getattr`. Could we use `hasattr` here?

Reply:
Changed to

```python
    try:
        return getattr(tp, name)
    except Exception as e:
        for base in reversed(mro(tp)):
            if hasattr(base, name):
                return getattr(base, name)
        raise e
```